### PR TITLE
Fix: use xpack.security.authc.realms.file

### PR DESCRIPTION
**Pull request by mtojek**
Issue: https://github.com/elastic/elastic-package/issues/234

### CODE REVIEW

**Comment by mtojek** [comment]
jenkins, run the tests please

**Comment by mtojek** [comment]
@ph I tried your suggestion, but I can't make it running. Most likely a conflict with the API key.

Logs: https://beats-ci.elastic.co/job/Ingest-manager/job/elastic-package/job/PR-240/3/artifact/build/elastic-stack-dump/stack/logs/elasticsearch.log

**Comment by ph** [comment]
@mtojek Interesting, I thought that the backend was an abstraction, seems not all backend are created equal. I do not have other ideas. Maybe sync with the ES team or @jen-huang maybe she knows.

File: internal/install/static_snapshot_yml.go [hunk 1]
```diff
@@ -21,11 +21,13 @@ services:
     - "xpack.license.self_generated.type=trial"
     - "xpack.security.enabled=true"
     - "xpack.security.authc.api_key.enabled=true"
+    - "xpack.security.authc.realms.file.file1.order=0"
     - "ELASTIC_PASSWORD=changeme"
     - "script.context.template.max_compilations_rate=unlimited"
     - "script.context.ingest.cache_max_size=2000"
     - "script.context.processor_conditional.cache_max_size=2000"
     - "script.context.template.cache_max_size=2000"
+    command: bash -c "bin/elasticsearch-users useradd elastic -r superuser -p changeme | /usr/local/bin/docker-entrypoint.sh eswrapper"
```
Comment (ycombinator) [inline internal/install/static_snapshot_yml.go:30]: Looks like the error is complaining about trying to add the `elastic` user which is a built-in user so it already exists. What happens if you remove this line?

Comment (mtojek) [reply internal/install/static_snapshot_yml.go:30]: I think that was a similar problem. I can give it a try.

Comment (mtojek) [reply internal/install/static_snapshot_yml.go]: Unfortunately same problem: https://beats-ci.elastic.co/job/Ingest-manager/job/elastic-package/job/PR-240/4/artifact/build/elastic-stack-dump/stack/logs/elasticsearch.log

**Comment by jen-huang** [comment]
> @mtojek Interesting, I thought that the backend was an abstraction, seems not all backend are created equal. I do not have other ideas. Maybe sync with the ES team or @jen-huang maybe she knows.

I see some interesting Kibana logs in that [last build](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-240/detail/PR-240/4/artifacts/) but I think I'm missing some context. Where can I find the test code that's executing the Kibana setup calls, etc?

**Comment by mtojek** [comment]
>I see some interesting Kibana logs in that last build but I think I'm missing some context. Where can I find the test code that's executing the Kibana setup calls, etc?

I believe it's just this file: https://github.com/elastic/beats/blob/master/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl

If you want to reproduce it, you can checkout this branch and:

```
make build
./elastic-package stack up -v -d
```

**Comment by mtojek** [comment]
Thank @jen-huang for taking a look. Let me know if you find anything. If not, I will raise it to the Elasticsearch team for some guidance.

**Comment by jen-huang** [comment]
> Thank @jen-huang for taking a look. Let me know if you find anything. If not, I will raise it to the Elasticsearch team for some guidance.

I was able to reproduce locally and while investigating I stumbled on an interesting issue: I'm not able to login with _any_ user other than `elastic` against the Elasticsearch instance that is spun up by `./elastic-package stack up -v -d`:

With elastic user, response is fine ✅
```
curl -X GET "localhost:9200" -u elastic:changeme
```

Create new user, user is created ✅
```
curl -X POST "localhost:9200/_security/user/testuser" -u elastic:changeme -H 'Content-Type: application/json' -d'
{
  "password" : "changeme",
  "roles" : [ "superuser" ]
}
'
```

Try to login with the new user, `unable to authenticate user` error is returned ❌
```
curl -X GET "localhost:9200" -u testuser:changeme
```

This makes me think the issue is indeed something to do with ES but I can't offer much insight there. @mtojek maybe you can see if this happens for you too?

**Comment by mtojek** [comment]
I checked this issue against the master branch (elastic-package):

```
curl -X GET "localhost:9200" -u testuser:changeme

{
  "name" : "ea3700e6cb7d",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "HdcGEZYsS6W00YQ2QLQT2Q",
  "version" : {
    "number" : "7.11.0-SNAPSHOT",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "6557e6ad7fbd149573dde6d866ca7b4c194c8a74",
    "build_date" : "2021-02-02T06:19:51.714374Z",
    "build_snapshot" : true,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

and it works for me.

PS. Maybe you have to update images: `elastic-package stack update -v`.

**Comment by mtojek** [comment]
Ok, reproduced your case @jen-huang with the PR branch:

```
curl -X GET "localhost:9200" -u testuser:changeme

{"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate user [testuser] for REST request [/]","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate user [testuser] for REST request [/]","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","ApiKey"]}},"status":401}%
```

**Comment by mtojek** [comment]
I moved a step forward. I had to create the user with:

```
$ elasticsearch-users useradd testuser2 -p changeme
$ elasticsearch-users roles testuser2 --add superuser
```

**Comment by mtojek** [comment]
It seems that with "users" file enabled the Elastic-Agent can't setup correctly:

```
{"isInitialized":true}{"statusCode":400,"error":"Bad Request","message":"Fleet Admin user is invalid: security_exception"}{"list":[],"total":0,"page":1,"perPage":20}
```

**Comment by mtojek** [comment]
I think I understand the situation. With enabled file-based authentication the Elastic Agent can't pass the setup because API keys are not supported?

```
➜  elastic-package git:(234-file-realm) ✗ docker logs 1f184ed75383 -f
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    22  100    22    0     0      1      0  0:00:22  0:00:11  0:00:11     5
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   100  100   100    0     0     93      0  0:00:01  0:00:01 --:--:--    93
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    43  100    43    0     0   1516      0 --:--:-- --:--:-- --:--:--  1535
{"isInitialized":true}{"statusCode":400,"error":"Bad Request","message":"Fleet Admin user is invalid: security_exception"}{"list":[],"total":0,"page":1,"perPage":20}
null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    82  100    82    0     0   2437      0 --:--:-- --:--:-- --:--:--  2484
null
The Elastic Agent is currently in BETA and should not be used in production

Error: fail to enroll: fail to execute request to Kibana: Status code: 401, Kibana returned an error: Unauthorized, message: [security_exception] missing authentication credentials for REST request [/_security/_authenticate], with { header={ WWW-Authenticate={ 0="Basic realm=\"security\" charset=\"UTF-8\"" & 1="ApiKey" } } }
``` 

I will open an issue for the Kibana/Agent to see if this is an issue we'd like to solve. Thank you, @jen-huang for hints!

**Comment by jen-huang** [comment]
@mtojek I should've noted I was testing against this PR branch 😄 Thanks for filing the Kibana issue. Here's some more context about what happens during [Fleet central management setup](https://github.com/elastic/beats/blob/master/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl#L17):

1. Setup creates the Fleet admin user (`fleet_enroll`)
2. Then it attempts to create an enrollment token (API key) for the Default policy by using the admin user's credentials
3. The admin user authentication fails due to the ES error we observed when using file-based auth, so the enrollment token isn't able to be created. This is the Kibana `Fleet Admin user is invalid` error. The setup endpoints returns `400` at this point - I agree with your note re: entry point should fail and exit here, rather than continuing to enroll.
4. Still, the entry point attempts to [retrieve an enrollment token to use](https://github.com/elastic/beats/blob/master/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl#L30)
5. Kibana will return none, but the entry point will still send another request to get [details of a token](https://github.com/elastic/beats/blob/master/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl#L43), but `$apiKeyId` value is `null`. This results in the ES `"java.lang.IllegalArgumentException: invalid ApiKey value"` error

**Pull request closed without merging**